### ISSUE TITLE
How to refactor synchronous custom binding patterns to work with deferred updates?

### ISSUE THREAD

**Issue by cervengoc** [knockout/knockout#1938]
Hello @mbest, as a follow-up of my previous post, I1m still working on trying to make our codes compatible with deferred updates.

At many custom bindings, where some editor or widget is wrapped inside I used to use a pattern to avoid circular updates. I mean something like this:

```
..
var componentUpdating = false;
var observable = valueAccessor();.
$(elem).on("changed", function(e) {
  componentUpdating = true;
  observable(e.value);
  componentUpdating = false;
});

ko.computed({
  read: function() {
    if (componentUpdating) return;
    $(elem).myFancyEditor("set", ko.unwrap(observable));
  },
  disposeWhenNodeIsRemoved: elem
});
...
```

This is a pattern which I saw at many places as a suggested technique to avoid circular updates between the editor changed event handler and the binding handler's computed.

Here one key thing is that the bound observable should have synchronous updates because if not, it will fall out of the protection.

What workaround or refactoring could you suggest for these cases?

**Comment by mbest** [comment]
The best method is to compare the values:

```
ko.computed({
  read: function() {
    if (elem.value !== observable())
        $(elem).myFancyEditor("set", observable());
  },
  disposeWhenNodeIsRemoved: elem
});
```

You also need to be careful when skipping parts of a computed's evaluator function, as you've done.  If `componentUpdating` is ever true in your example, the computed will be become a "zombie" because it will have no dependencies. See http://knockoutjs.com/documentation/computed-dependency-tracking.html

**Comment by mbest** [comment]
If it's difficult to compare the element value, or if the value stored in the observable is an object, you could possibly save the value for comparison:

```
var componentWrittenValue;
var observable = valueAccessor();.
$(elem).on("changed", function(e) {
  observable(componentWrittenValue = e.value);
});

ko.computed({
  read: function() {
    if (observable() !== componentWrittenValue)
        $(elem).myFancyEditor("set", observable());
  },
  disposeWhenNodeIsRemoved: elem
});
```

**Comment by mbest** [comment]
Knockout's built-in bindings tend to do this kind of check when writing to the observable instead of when writing to the view.

```
$(elem).on("changed", function(e) {
  if (observable.peek() !== e.value) {
    observable(e.value);
  }
});
```

**Comment by cervengoc** [comment]
Thank you, you wrote some great ideas and best practices.

Although I know about it, it was late at night when I posted it, and forgot to care about avoiding "zombiness".

I also came to an idea after I posted the question. If I can be sure that the external widget will fire the `changed` event synchronously, then I can simply add the flag handling to the change handler and kind of reverse the logic. Am I right?
